game/src/engine/pickup/PickupService.h

```diff
@@ -0,0 +1,120 @@
+                                        const std::string& member);
+
+    // Only for buckshot as it has other information as well.
+    void LoadAmmoOtherInformation(const rapidjson::Value& ammo_object,
+                                  const std::string& member);
+
+    /******** POWERUPS ******/
+    void LoadPowerupInformation(const rapidjson::Document& doc);
+
+    void LoadPowerupDurationInformation(const rapidjson::Value& powerup_object,
+                                        const std::string& member);
+
+    void LoadPowerupOtherInformation(const rapidjson::Value& powerup_object,
+                                     const std::string& member);
+    void LoadPowerupRespawnTimeInformation(
+        const rapidjson::Value& powerup_object, const std::string& member);
+
+    /****** GETTERS SETTERS *****/
+    std::array<std::string, 6> GetAmmoPickupNames();
+    float GetAmmoDamage(std::string ammo_type);
+    float GetAmmoDuration(std::string ammo_type);
+    float GetAmmoCooldown(std::string ammo_type);
+    float GetAmmoRespawnTime(std::string ammo_type);
+    float GetBuckshotAdditionalDetail(std::string detail_type);
+
+    std::array<std::string, 5> GetPowerupPickupNames();
+    float GetPowerupDuration(std::string powerup_type);
+    float GetPowerupMaxSpeeds(std::string powerup_type);
+    float GetPowerupRespawnTime(std::string powerup_type);
+
+    /********* EXECUTING AND HANDLING THE POWERUPS ************/
+
+    // Handle disabling the powerup
+    void HandleDisablingPowerup();
+
+    // Add the entity with the powerup they are holding
+    void AddEntityWithPowerup(Entity* entity, const std::string& powerup);
+
+    // Add the timers for the powerup when they are executed.
+    void AddEntityWithTimer(Entity* entity, const float& timer);
+
+    // To tell if the car should have the slowing down multiplier to them or not
+    bool IsVehicleSlowDown(Entity* entity);
+
+    // To tell if the car should have the slowing down multiplier to them or not
+    bool IsVehicleDisableHandling(Entity* entity);
+
+  private:
+    /******** AMMO ********/
+    std::unordered_map<std::string, float> ammo_damages_;
+    std::unordered_map<std::string, float> ammo_durations_;
+    std::unordered_map<std::string, float> ammo_cooldowns_;
+    std::unordered_map<std::string, float> buckshot_additional_details_;
+    std::unordered_map<std::string, float> ammo_respawn_times_;
+
+    /******** POWERUP ***/
+    std::unordered_map<std::string, float> powerup_durations_;
+    std::unordered_map<std::string, float> powerup_max_speeds_;
+    std::unordered_map<std::string, float> powerup_respawn_times_;
+
+    /********* EXECUTING AND HANDLING THE POWERUPS ************/
+
+    // Storing the {entity, powerup}
+    std::unordered_map<Entity*, std::string> entity_holding_powerups_;
+
+    // Storing the {entity, timer}
+    std::unordered_map<Entity*, float> entity_timer_powerups_;
```

Comment:
Rather than having two separate maps, I think it might be useful to have one single map that goes from Entity* -> (some struct) instead. This would likely make it much easier to manage, since it's technically possible to have the two maps get out of sync otherwise. It'd also make `HandleDisablingPowerup` simpler too, since you could do like:

```c++
struct PlayerPowerupState
{
    std::string powerup_name;
    float timer;
}
void HandleDisablingPowerup()
{
    auto iter = entity_powerups_.begin();
    while (iter != entity_powerups_.end())
    {
        float max_time = ammo_info_[iter->second.powerup_name];
        if (iter->second.timer > max_time)
        {
              // i believe that once you erase an item, an iterator is no longer valid, so you'd need to
              // break after if you do it this way
              entity_powerups_.erase(iter);
              break;
        }
        iter++
    }
}
```

Reply:
This will take too much time as I would need to change some stuff I am doing in every powerup as well. There will never be more than 2 maps anyways, can we skip this :( .

Reply:
No worries, all good!